Updated Clever Girl to consider all available ammo and modes

redbatz opened the pull request:
Updated Clever Girl to consider all available ammo and modes  when deciding which weapons to fire.

First step, more tuning needed as this change exposed issues with the AI not understanding a number of key stats.

redbatz committed:
Updated Clever Girl to consider all available ammo and modes when deciding which weapons to fire.
First step, more tuning needed as this change exposed issues with the AI not understanding a number of key stats.

IceRaptor commented on CleverGirl/CleverGirl/Helper/AOHelper.cs:
```diff
@@ -123,16 +134,16 @@ public static class AOHelper {
                 {
                     weaponListSB.Append("'");
-                    weaponListSB.Append(weapon3.Name);
+                    weaponListSB.Append(wamp.Key.UIName);
```
Can this throw a null, if the wamp.key isn't found?

redbatz replied:
Should never be null, but adding a null-check.

IceRaptor commented on CleverGirl/CleverGirl/Helper/AOHelper.cs:
```diff
@@ -457,7 +489,16 @@ private static bool DoesMechFavorMelee(Mech attacker)
+                        foreach (AmmoModePair ammoModePair in weapon.getAvaibleFiringMethods())
+                        {
+                            weapon.ApplyAmmoMode(ammoModePair);
+                            optimalDamage = Mathf.Max(optimalDamage, (int)(weapon.DamagePerShot * weapon.ShotsWhenFired));
```
Why are you doing a floating point compare here, when you've cast to int and optimalDamage is 0?

IceRaptor replied:
Additionally, this calc isn't taking quite a bit of CAC logic into account. It's ignoring AoE fragmentation, AoE falloff, etc. CG probably ignored that as well, so more just a caution since I'm not entirely sure what this block of checking is intended to achieve.

redbatz replied:
It's a very naive approach to trying to find if the mech should consider melee attacks. I did not change it from existing code besides adding ammomodes. 
Adding a todo to look at using the CAC logic, but there's a performance impact of that I would imagine.

IceRaptor commented on CleverGirl/CleverGirl/Helper/BehaviorHelper.cs:
```diff
@@ -17,23 +20,43 @@ public static BehaviorVariableValue GetBehaviorVariableValue(BehaviorTree bTree,
+                case BehaviorVariableValue.BehaviorVariableType.String:
+                    return behaviorVariableValue.StringVal;
+                default:
+                    throw new ArgumentOutOfRangeException();
```
Throwing an exception here will cause the entire logic to brick (because it's not caught elsewhere), which I think should cause an AI unit to brace. If that's intentional great, but might be better to return an empty string?

IceRaptor commented on CleverGirl/CleverGirl/Objects/AmmoModeAttackEvaluation.cs:
```diff
@@ -0,0 +1,34 @@
+    public float ExpectedDamage;
+    public float lowestHitChance;
+
+    public int CompareTo(object otherObj)
```
Are you dropping the utility comparison entirely in this?

redbatz replied:
Not sure what you mean? The compare to is used to sort the list of attacks later, which I am not sure why it was done as it iterated over the whole list later anyway, but I kept it.

IceRaptor replied:
I mean the 'utility' like heat-as-damage, knockdown, etc. It's present at https://github.com/BattletechModders/CleverGirl/blob/11d38af111c0ced70f3e094a3993524502d6a945/CleverGirl/CleverGirl/Helper/AIHelper.cs#L252 when calculating in the EV in the main branch. Wasn't sure from my reading how your changes would intersect with that.

IceRaptor commented on CleverGirl/CleverGirl/Objects/CandidateWeapons.cs:
```diff
@@ -69,37 +67,44 @@ public CandidateWeapons(AbstractActor attacker, ICombatant target)
                     {
-                        rawWeapon.ApplyAmmoMode(item);
+                        rawWeapon.ApplyAmmoMode(ammoModePair);
                         bool modeWillFire = rawWeapon.WillFireAtTargetFromPosition(target, attacker.CurrentPosition, attacker.CurrentRotation);
```
IIRC, using attacker.CurrentPosition is a bug in current CG. Since you're changing the surrounding logic, you might want to change this to the attack position instead.

redbatz replied:
Not sure what attack position would be otherwise, we don't know any movement here do we? All the logic uses CurrentPosition as far as I can find.

IceRaptor replied:
Eh... kinda? The BehTree makes this weird. As I understanding it (which could very well be wrong!) the target position for movement is selected an attack order is evaluated. Except in the case of melee, in which case the best melee/DFA position is calculated at https://github.com/BattletechModders/CleverGirl/blob/11d38af111c0ced70f3e094a3993524502d6a945/CleverGirl/CleverGirl/Helper/AOHelper.cs#L66. When evaluating weapons attacks it can be either 

1) !isStationary, so ranged attack and presumably currentPosition is fine
2) isStationary, ranged attack, presumably currentPosition is fine
3) isStation, melee/DFA attack, need to use best melee position

I could be overthinking this though.

IceRaptor commented on CleverGirl/CleverGirl/Objects/CandidateWeapons.cs:
```diff
@@ -69,37 +67,44 @@ public CandidateWeapons(AbstractActor attacker, ICombatant target)
+                            cWeapon.AddAmmoMode(ammoModePair);
                         }
                     }
+                    rawWeapon.ApplyAmmoMode(currentAmmoMode);
```
Nice restoring to a known good value.

IceRaptor commented on CleverGirl/CleverGirl/Objects/CandidateWeapons.cs:
```diff
@@ -113,8 +118,25 @@ public CandidateWeapons(AbstractActor attacker, ICombatant target)
+            foreach (AmmoModePair ammoModePair in weapon.getAvaibleFiringMethods())
+            {
+                weapon.ApplyAmmoMode(ammoModePair);
+                if (weapon.CanFire)
```
IIRC I think weapon.CanFire already checks the ammo count, but if not might be worth checking here?

redbatz replied:
It does check the ammo count through a CAC transpile on HasAmmo inside the call to check the set ammomodes count, that is why I switch the ammomodes around.

IceRaptor replied:
Great

IceRaptor commented on CleverGirl/CleverGirl/Objects/CleverGirlComponent.cs:
Did you remove this because it's unused? Good if you did so!

IceRaptor commented on CleverGirl/CleverGirl/Helper/AIHelper.cs:
```diff
@@ -53,16 +59,39 @@ public class AIHelper {
 
             float blowQualityMultiplier = unit.Combat.ToHit.GetBlowQualityMultiplier(attackParams.Quality);
-            float totalDam = totalExpectedDam * blowQualityMultiplier;
+            float artilleryDamage = totalExpectedDam.GetValueOrDefault(true, 0f) * blowQualityMultiplier;
```
FWIW I think using blowQualityMultiplier on my end might be a carry over from HBS, that's not actually used? I vaguely recall it being used for the 'glancing hit' mechanic, but I think it's mostly a 1f at this point?

redbatz replied:
From what I can see it actually does check brace, sensorlock etc to find a guard level of the target, which translate to blow quality.

IceRaptor commented on CleverGirl/CleverGirl/Helper/AEHelper.cs:
```diff
@@ -212,295 +177,87 @@ public static bool MeleeDamageOutweighsRisk(float attackerMeleeDam, Mech attacke
-                    float toHitFromPosition = weapon.First.GetToHitFromPosition(target, 1, shooterPosition, target.CurrentPosition, true, true, false);
-                    if (toHitFromPosition < toHitFrac)
+                    // Check one-shot weapons for accuracy
+                    if (wep.weaponDef.StartingAmmoCapacity == wep.weaponDef.ShotsWhenFired)
```
I believe BTA has a 'reloading' One-Shot weapons, which is why the prior logic existed. You should check with BD that this change will continue to account for those weapons.

redbatz replied:
Discussed it with BTA crew and decided to switch to a more in-depth check using CAC features since `StartingAmmoCapacity` is not reliable as definition of internal ammunition.

IceRaptor commented on CleverGirl/CleverGirl/Helper/AEHelper.cs:
```diff
@@ -212,295 +177,87 @@ public static bool MeleeDamageOutweighsRisk(float attackerMeleeDam, Mech attacke
         {
-
-            Mech mech = target as Mech;
-            if (mech == null || !mech.IsVulnerableToCalledShots() || evaluatedAttack.AttackType == AIUtil.AttackType.Melee || evaluatedAttack.AttackType == AIUtil.AttackType.DeathFromAbove)
```
Are you reverting to the basic logic for all of this, and it's just not obvious from the diff? Or does this move elsewhere?

redbatz replied:
None of the clone HBS code was actually ever called besides MakeWeaponSets which I replaced. The other code was only called from commented out code like the fractioncal evasion checking or these [comments](https://github.com/BattletechModders/CleverGirl/blob/11d38af111c0ced70f3e094a3993524502d6a945/CleverGirl/CleverGirl/Helper/AOHelper.cs#L257) in AOHelper .

I am not sure what your intentions were here, to eventually reenabled it or if it was removed for performance reasons.

IceRaptor replied:
I was probably just dumb

IceRaptor reviewed (commented):
Overall looks solid, there is a surprising amount of code deletion (mostly HBS clones). So long as those are intention, should be fine.

redbatz committed:
Fixes

redbatz committed:
Updates to One-Shot logic

redbatz committed:
Optimizations
- Now uses a cache for predicted damage to reduce need to recalculate same CondensedWeapon/AmmoMode several times.
- If multiple AttackEvaluations return the exact same values, just keep the original and discrd the rest.

tabder committed:
wip#1
